### PR TITLE
Release/0.10

### PR DESCRIPTION
**Pull request by thefosk**

### CODE REVIEW

File: app/docs/0.10.x/proxy.md [hunk 1]
```diff
@@ -0,0 +1,759 @@
+---
+title: Proxy Reference
+---
+
+# Proxy Reference
+
+As you may already know, Kong listens for traffic on three ports, which by
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: Suggesting to remove:

> As you may already know,

File: app/docs/0.10.x/proxy.md [hunk 2]
```diff
@@ -0,0 +1,759 @@
...
+default are:
+
+- `:8001` on which the [Admin API][API] used to configure Kong listens.
+- `:8000` on which Kong listens for incoming HTTP traffic from your clients, and 
+forwards it to your upstream services. This is the port that interests us in 
+this guide.
+- `:8443` on which Kong listens for incoming HTTPs traffic. This port has a
+similar behavior as the `:8000` port, except that it expects HTTPs traffic. This
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: `HTTPs` should be `HTTPS`

File: app/docs/0.10.x/proxy.md [hunk 3]
```diff
@@ -0,0 +1,759 @@
...
+port can be disabled via the configuration file.
+
+We will cover routing capabilities of Kong by explaining in details how incoming
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: `details` should be `detail`

File: app/docs/0.10.x/proxy.md [hunk 4]
```diff
@@ -0,0 +1,759 @@
...
+this guide.
+- `:8443` on which Kong listens for incoming HTTPs traffic. This port has a
+similar behavior as the `:8000` port, except that it expects HTTPs traffic. This
+port can be disabled via the configuration file.
+
+We will cover routing capabilities of Kong by explaining in details how incoming 
+requests on port `:8000` are proxied to a configured upstream service depending
+on their headers, URI, and HTTP method.
+
+It is worth pointing out that both of these ports can be configured. See the
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: Both implies two, however I see three ports. Clarification required.

File: app/docs/0.10.x/proxy.md [hunk 5]
```diff
@@ -0,0 +1,759 @@
...
+[Back to TOC](#table-of-contents)
+
+### Overview
+
+From a high level perspective, Kong will listen for HTTP traffic on its
+configured proxy port (`8000` by default), recognize which upstream service is
+being requested, run the configured plugins for that API, and forward the HTTP 
+request upstream to your own API or service. Kong can also run plugins once the 
+response from your upstream service is received, but that topic is not as
+relevant to this guide.
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: > Kong can also run plugins once the response from your upstream service is received, but that topic is not as relevant to this guide.

Where do I find how to do this? Needs reference, otherwise remove.

File: app/docs/0.10.x/proxy.md [hunk 6]
```diff
@@ -0,0 +1,759 @@
...
+
+When a client makes a request to the proxy port, Kong will decide where to route
+(or forward) that incoming request (that is, to which of your upstream
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: `that` should be `the`

File: app/docs/0.10.x/proxy.md [hunk 7]
```diff
@@ -0,0 +1,759 @@
...
+response from your upstream service is received, but that topic is not as
+relevant to this guide.
+
+When a client makes a request to the proxy port, Kong will decide where to route
+(or forward) that incoming request (that is, to which of your upstream 
+services). Kong achieves this through the APIs you configured (via the Admin 
+API). You can configure APIs with various properties, but the three relevant 
+ones for routing incoming traffic are `hosts`, `uris` and `methods`.
+
+If Kong cannot determine to which API a given request should be routed to, it
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: can remove `to` in `determine to which` it is redundant

File: app/docs/0.10.x/proxy.md [hunk 8]
```diff
@@ -0,0 +1,759 @@
...
+Server: kong/<x.x.x>
+
+{
+    "message": "no API found with those values"
+}
+```
+
+### Reminder: how to add an API to Kong
+
+Before goign any further, let's take a few moments to make sure you know how to
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: `goign` should be `going`

File: app/docs/0.10.x/proxy.md [hunk 9]
```diff
@@ -0,0 +1,759 @@
...
+This request instructs Kong to register an API named "my-api", reachable at
+"http://my-api.com". It also specifies various routing properties. In reality,
+**only one of** `hosts`, `uris` and `methods` is actually required. More on this
+later.
+
+Adding such an API would mean that you configured Kong to proxy all incoming 
+requests matching `hosts`, `uris`, and `methods` to `http://my-api.com`.
+Kong being a transparent proxy, it will forward the request to your upstream
+service - practically - untouched (exception made of various headers such as
+`Connection`, or in case a plugin performs some transformations to it).
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: We can remove the `- practically -` portion, it is implied in the clarification after `untouched`

`exception made of` should be `with the exception of`

File: app/docs/0.10.x/proxy.md [hunk 10]
```diff
@@ -0,0 +1,759 @@
...
+    -d 'hosts=my-api.com' \
+    -d 'uris=/my-api' \
+    -d 'methods=GET,HEAD'
+HTTP/1.1 201 Created
+...
+```
+
+This request instructs Kong to register an API named "my-api", reachable at
+"http://my-api.com". It also specifies various routing properties. In reality,
+**only one of** `hosts`, `uris` and `methods` is actually required. More on this
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: `actually` can be removed, it is not necessary here

File: app/docs/0.10.x/proxy.md [hunk 11]
```diff
@@ -0,0 +1,759 @@
...
+requests matching `hosts`, `uris`, and `methods` to `http://my-api.com`.
+Kong being a transparent proxy, it will forward the request to your upstream
+service - practically - untouched (exception made of various headers such as
+`Connection`, or in case a plugin performs some transformations to it).
+
+[Back to TOC](#table-of-contents)
+
+### Routing capabilities
+
+Let's now investigate in details how Kong matches a request to the configured
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: `details` should be `detail`

File: app/docs/0.10.x/proxy.md [hunk 12]
```diff
@@ -0,0 +1,759 @@
...
+`hosts`, `uris` and `methods` properties of your API:
+
+- All of those fields are **optional**, but at least **one of them** must be 
+specified.
+- For a request to match an API, all of configured fields **must** be satisfied
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: `all of configured` should be `all of the configured`

File: app/docs/0.10.x/proxy.md [hunk 13]
```diff
@@ -0,0 +1,759 @@
...
+(present in the client HTTP request).
+- All of those fields accept one or more values. A field only needs one of
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: `All of those` should be `All of these` as we are currently in the context of these fields.

File: app/docs/0.10.x/proxy.md [hunk 14]
```diff
@@ -0,0 +1,759 @@
...
+    -d 'upstream_url=http://my-api.com' \
+    -d 'hosts=my-api.com' \
+    -d 'uris=/my-api' \
+    -d 'methods=GET,HEAD'
+HTTP/1.1 201 Created
+...
+```
+
+This request instructs Kong to register an API named "my-api", reachable at
+"http://my-api.com". It also specifies various routing properties. In reality,
```
Comment (p0pr0ck5) [inline app/docs/0.10.x/proxy.md]: It is idiomatic to use 'example.com' (or some thing RFC2606/6761 compliant hostname/tld) for cases such as this. Consider renaming cases of 'my-api.com' to 'example.com' or 'my-api.example'.

File: app/docs/0.10.x/proxy.md [hunk 15]
```diff
@@ -0,0 +1,759 @@
...
+GET /bar HTTP/1.1
+Host: service.com
+```
+
+```http
+GET /foo/hello/world HTTP/1.1
+Host: example.com
+```
+
+All three of those requests satisfy all of the three conditions set in the API
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: Instead of repeating `three` we can consolidate the sentence  to:

> All three of these requests satisfy all the conditions set in the API definition.

File: app/docs/0.10.x/proxy.md (hunk 1 above)
Comment (coopr) [inline app/docs/0.10.x/proxy.md]: Agree

File: app/docs/0.10.x/proxy.md [hunk 16]
```diff
@@ -0,0 +1,759 @@
...
+
+# Proxy Reference
+
+As you may already know, Kong listens for traffic on three ports, which by
+default are:
+
+- `:8001` on which the [Admin API][API] used to configure Kong listens.
+- `:8000` on which Kong listens for incoming HTTP traffic from your clients, and 
+forwards it to your upstream services. This is the port that interests us in 
+this guide.
```
Comment (coopr) [inline app/docs/0.10.x/proxy.md]: *This is the port that interests us in this guide.*

Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: I think @coopr is implying it should be made bold

File: app/docs/0.10.x/proxy.md [hunk 17]
```diff
@@ -0,0 +1,759 @@
...
+
+```http
+GET /foo/hello/world HTTP/1.1
+Host: example.com
+```
+
+All three of those requests satisfy all of the three conditions set in the API
+definition.
+
+However, the following requests would **not** match this API:
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: We can omit `this API` here

File: app/docs/0.10.x/proxy.md [hunk 18]
```diff
@@ -0,0 +1,759 @@
...
+understanding about how Kong forwards incoming requests to your upstream 
+services, now is a good opportunity to spend some time detailing what happens 
+internally between the time Kong *recognized* an HTTP request to target a 
+service, and actual *forwarding* of the request to it.
+
+[Back to TOC](#table-of-contents)
+
+#### 1. Load balancing
+
+Sarting with Kong 0.10, Kong implements load balancing capabilities, allowing it
```
Comment (p0pr0ck5) [inline app/docs/0.10.x/proxy.md]: typo (sarting => starting)

File: app/docs/0.10.x/proxy.md [hunk 19]
```diff
@@ -0,0 +1,759 @@
...
+POST /foo HTTP/1.1
+Host: example.com
+```
+
+```http
+GET /foo HTTP/1.1
+Host: foo.com
+```
+
+All three of those requests are only satisfying two out of the three configured
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: `those` should be `these`

`only satisfying` should be `satisfying only`

`out of the three` should be `of three`

File: app/docs/0.10.x/proxy.md [hunk 20]
```diff
@@ -0,0 +1,759 @@
...
+conditions. The first request's URI is not a match for any of the configured 
+`uris`, and it is so for the second request's HTTP method, or the third's Host
+header.
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: >  and it is so for the second request's HTTP method, or the third's Host header.

would read better as something like

> same for the second request's HTTP `method`, and the third request's `Host` header.

File: app/docs/0.10.x/proxy.md [hunk 21]
```diff
@@ -0,0 +1,759 @@
...
+### Terminology
+
+In this document, we will refer to various Kong concepts. For the sake of 
+consistency, we will now define terms refering to such concepts:
+
+- `API`: This term refers to the API entity of Kong. You configure your APIs,
+that point to your own Upstream services through the Admin API.
+- `Plugin`: This refers to Kong "plugins", which are pieces of business logic
+that run in proxying lifecycle. Plugins can be configured globally (all 
+incoming traffic), or on a per-API basis through the Admin API.
```
Comment (coopr) [inline app/docs/0.10.x/proxy.md]: Plugins can be configured through the Admin API - either globally (all incoming traffic) or on a per-API and/or per-Consumer basis.

File: app/docs/0.10.x/proxy.md [hunk 22]
```diff
@@ -0,0 +1,759 @@
...
+granularity when configuring APIs and Plugins. For example, one could imagine 
+two APIs pointing to the same upstream service: one allowing for `GET` requests
+and unprotected, and a second one filtering out `POST` requests and applying 
+authentication or rate limiting plugins on such requests.
+
+[Back to TOC](#table-of-contents)
+
+### Routing priorities
+
+It is now clear that an API may define matching rules based on its `hosts`,
```
Comment (p0pr0ck5) [inline app/docs/0.10.x/proxy.md]: "It is now clear" sentences seem unnecessary. If it's necessary to restate these conditions, let's reword this to not use a condescending tone. Otherwise, we can scrap these two sentences.

File: app/docs/0.10.x/index.md [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+  <div class="docs-grid-block">
+    <h3><img src="/assets/images/icons/documentation/icn-doc-reference.svg" /><a href="/docs/{{page.kong_version}}/proxy">Proxy reference</a></h3>
+    <p>Learn every way to configure Kong to proxy your APIs, serve them over SSL or use WebSockets.</p>
+    <a href="/docs/{{page.kong_version}}/proxy">Read the Proxy Reference &rarr;</a>
+  </div>
+
+  <div class="docs-grid-block">
+    <h3><img src="/assets/images/icons/documentation/icn-doc-reference.svg" /><a href="/docs/{{page.kong_version}}/loadbalancing">Load balancing reference</a></h3>
+    <p>Learn how to setup Kong to load balance traffic through replicas of your upstream services.</p>
+    <a href="/docs/{{page.kong_version}}/proxy">Read the Load balancing Reference &rarr;</a>
```
Comment (p0pr0ck5) [inline app/docs/0.10.x/index.md]: inconsistent capitalization in 'Load balancing Reference"; either capitalize Balancing, or lowercase load and reference

File: app/docs/0.10.x/proxy.md [hunk 24]
```diff
@@ -0,0 +1,759 @@
...
+Host: foo.com
+```
+
+All three of those requests are only satisfying two out of the three configured
+conditions. The first request's URI is not a match for any of the configured 
+`uris`, and it is so for the second request's HTTP method, or the third's Host
+header.
+
+Now that we grasp how each of those three properties play together, let's go
+into more details and explore each one of them individually.
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: `those` should be `these` and `properties` should be `fields`

File: app/docs/0.10.x/proxy.md (hunk 22 above)
Comment (thibaultcha) [inline app/docs/0.10.x/proxy.md]: One of:

"It should now be clear"
"It should now be established"
"Having established an API may define [...]"

?

File: app/docs/0.10.x/proxy.md [hunk 25]
```diff
@@ -0,0 +1,759 @@
...
+    - [Request HTTP method][proxy-request-http-method]
+- [Routing priorities][proxy-routing-priorities]
+- [Proxying behavior][proxy-proxying-behavior]
+    - [1. Load balancing][proxy-load-balancing]
+    - [2. Plugins execution][proxy-plugins-execution]
+    - [3. Proxying & upstream timeouts][proxy-proxying-upstream-timeouts]
+    - [4. Response][proxy-response]
+- [Configuring a fallback API][proxy-configuring-a-fallback-api]
+- [Configuring SSL for an API][proxy-configuring-ssl-for-an-api]
+    - [The `https_only` property][proxy-the-https-only-property]
```
Comment (thibaultcha) [inline app/docs/0.10.x/proxy.md]: link doesn't seem to work

File: app/docs/0.10.x/proxy.md [hunk 26]
```diff
@@ -0,0 +1,759 @@
...
+    - [The `http_if_terminated` property][proxy-the-http-if-terminated-property]
```
Comment (thibaultcha) [inline app/docs/0.10.x/proxy.md]: link doesn't seem to work

File: app/docs/0.10.x/proxy.md (hunk 14 above)
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: Agreed, as later on we use `example.com`

File: app/docs/0.10.x/proxy.md (hunk 22 above)
Comment (p0pr0ck5) [inline app/docs/0.10.x/proxy.md]: The third option feels the most palatable.

File: app/docs/0.10.x/proxy.md [hunk 27]
```diff
@@ -0,0 +1,759 @@
...
+- `:8443` on which Kong listens for incoming HTTPs traffic. This port has a
+similar behavior as the `:8000` port, except that it expects HTTPs traffic. This
+port can be disabled via the configuration file.
+
+We will cover routing capabilities of Kong by explaining in details how incoming 
+requests on port `:8000` are proxied to a configured upstream service depending
+on their headers, URI, and HTTP method.
+
+It is worth pointing out that both of these ports can be configured. See the
+[Configuration Reference] for more details.
```
Comment (hutchic) [inline app/docs/0.10.x/proxy.md]: missing link?

File: app/docs/0.10.x/proxy.md [hunk 28]
```diff
@@ -0,0 +1,759 @@
...
+```
+Host: service.com
+```
+
+[Back to TOC](#table-of-contents)
+
+##### Using wildcard hostnames
+
+To provide flexibility, Kong allows you to specify hostnames with wildcards in
+the `hosts` field. Such hostnames allow any matching Host header to satisfy the
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: `Such hostnames` should be `Wildcard hostnames`

File: app/docs/0.10.x/proxy.md [hunk 29]
```diff
@@ -0,0 +1,759 @@
...
+condition, and match a given API. 
+
+Such hostnames **must** contain **only one** asterisk at the leftmost **or**
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: `Such hostnames` should be `Wildcard hostnames`

File: app/docs/0.10.x/proxy.md [hunk 30]
```diff
@@ -0,0 +1,759 @@
...
+
+```http
+GET / HTTP/1.1
+Host: my-api.com
+```
+
+The Host header value is extracted from the hostname of the API's `upstream_url`
+field.
+
+However, by configuring your API with `preserve_host=true`:
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: Should further clarify what adding this will do in the sentence, either before the examples in this sentence, or in the next sentence. It is currently relying on the example to imply it's action.

File: app/docs/0.10.x/proxy.md [hunk 31]
```diff
@@ -0,0 +1,759 @@
...
+```http
+GET / HTTP/1.1
+Host: service.com
+```
+
+[Back to TOC](#table-of-contents)
+
+#### Request URI
+
+Another way for Kong to route a request to a given upstream service of yours is
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: - `of yours`

File: app/docs/0.10.x/proxy.md [hunk 32]
```diff
@@ -0,0 +1,759 @@
...
+GET /service/resource?param=value HTTP/1.1
+Host: my-api.com
+```
+
+```http
+GET /hello/world/resource HTTP/1.1
+Host: anything.com
+```
+
+For each of those requests, Kong detects that their URI is prefixed with one of
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: `those` should be `these`

File: app/docs/0.10.x/proxy.md [hunk 33]
```diff
@@ -0,0 +1,759 @@
...
+the API's `uris` values. By default, Kong would then forward the request 
+upstream with the untouched, **same URI**.
+
+When proxying with URIs prefixes, **the longest URIs get evaluated first**. This
+allow you to define two APIs with two URIs: `/service` and `/service/resource`,
+and ensure that the former does not "shadow" the later.
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: `later` should be `latter`

File: app/docs/0.10.x/proxy.md [hunk 34]
```diff
@@ -0,0 +1,759 @@
...
+```
+
+Will make Kong send the following request to your upstream service:
+
+```http
+GET /path/to/resource HTTP/1.1
+Host: my-api.com
+```
+
+This property can come very handy at times.
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: `come` should be `be`

File: app/docs/0.10.x/proxy.md [hunk 35]
```diff
@@ -0,0 +1,759 @@
...
+GET / HTTP/1.1
+Host:
+```
+
+```http
+HEAD /resource HTTP/1.1
+Host:
+```
+
+But would not match a `POST` or a `DELETE` request. This allows for much more
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: can remove `a` in `or a` here

File: app/docs/0.10.x/proxy.md [hunk 36]
```diff
@@ -0,0 +1,759 @@
...
+[proxy-the-http-if-terminated-property]: #the-http_if_terminated-property
+[proxy-websocket]: #proxy-websocket-traffic
+[proxy-conclusion]: #conclusion
+
+---
+
+### Terminology
+
+In this document, we will refer to various Kong concepts. For the sake of 
+consistency, we will now define terms refering to such concepts:
```
Comment (hutchic) [inline app/docs/0.10.x/proxy.md]: s/refering/referring

File: app/docs/0.10.x/proxy.md [hunk 37]
```diff
@@ -0,0 +1,759 @@
...
+
+```http
+HEAD /resource HTTP/1.1
+Host:
+```
+
+But would not match a `POST` or a `DELETE` request. This allows for much more 
+granularity when configuring APIs and Plugins. For example, one could imagine 
+two APIs pointing to the same upstream service: one allowing for `GET` requests
+and unprotected, and a second one filtering out `POST` requests and applying
```
Comment (nijikokun) [inline app/docs/0.10.x/proxy.md]: > one allowing for `GET` and unprotected requests

File: app/docs/0.10.x/proxy.md [hunk 38]
```diff
@@ -0,0 +1,759 @@
...
+services, now is a good opportunity to spend some time detailing what happens 
+internally between the time Kong *recognized* an HTTP request to target a 
+service, and actual *forwarding* of the request to it.
+
+[Back to TOC](#table-of-contents)
+
+#### 1. Load balancing
+
+Sarting with Kong 0.10, Kong implements load balancing capabilities, allowing it
+to distribute the forwarded requests accrodd replicas of your upstream services.
```
Comment (hutchic) [inline app/docs/0.10.x/proxy.md]: accrodd

File: app/docs/0.10.x/proxy.md [hunk 39]
```diff
@@ -0,0 +1,759 @@
...
+- `X-Real-IP: <$proxy_add_x_forwarded_for>`, where `$proxy_add_x_forwarded_for`
+is the variable bearing the same name provided by
+[ngx_http_proxy_module][ngx-http-proxy-module].
+- `X-Forwarded-Proto: <protocol>`, where `<protocol>` is the protocol used by 
+the client.
+
+All the other request headers are forwarded as-is by Kong.
+
+One exception to this is made when using the WebSocket protocol. If so, Kong
+will set the following headers to allow for upgrading the protocol bewteen the
```
Comment (hutchic) [inline app/docs/0.10.x/proxy.md]: between

File: app/docs/0.10.x/proxy.md [hunk 40]
```diff
@@ -0,0 +1,759 @@
...
+taken by Kong between receiving the request from the client and sending the 
+request to your upstream service.
+- `Kong-Upstream-Latency: <latency>`, where `latency` is the time in 
+milliseconds spent by Kong waiting for the first byte of your upstream service
+response.
+
+Once the headers are sent back to the client, Kong will start executing 
+registered plugins for that API that implement the `body_filter` hook. This hook
+may be called multiple times, due to the streaming nature of Nginx itself. Each
+chunk of the upstream response that is successfuly processed by such
```
Comment (hutchic) [inline app/docs/0.10.x/proxy.md]: s/successfuly/successfully

File: app/docs/0.10.x/proxy.md [hunk 41]
```diff
@@ -0,0 +1,759 @@
...
+This will make Kong forward the `Connection` and `Upgrade` headers to your
+upstream service, instead of dismissing them due to the hop-by-hop nature of a
+standard HTTP proxy.
+
+[Back to TOC](#table-of-contents)
+
+### Conclusion
+
+Through this guide, we hope you gained knowledge of the underlying proxying
+mehanism of Kong, from how is a request matched to an API, to how to allow for
```
Comment (hutchic) [inline app/docs/0.10.x/proxy.md]: s/mehanism/mechanism

File: app/docs/0.10.x/proxy.md (hunk 22 above)
Comment (coopr) [inline app/docs/0.10.x/proxy.md]: In my editing pass, I'm stripping out most of these "connector phrases" as they assume the document is being read through from top to bottom. While some users may read it that way, others will not.
